ambry-store/src/main/java/com.github.ambry.store/LogSegment.java

```diff
@@ -0,0 +1,391 @@
+   * @param size The amount of data in bytes to be written from the channel
+   * @throws IllegalArgumentException if there is not enough space for {@code size} data and forwarding is not possible
+   * @throws IOException if data could not be written to the file because of I/O errors
+   * @throws UnsupportedOperationException if this LogSegment is intended to be swap space
+   */
+  @Override
+  public void appendFrom(ReadableByteChannel channel, long size)
+      throws IOException {
+    if (state.equals(State.SEALED)) {
+      ensureNext(metrics.overflowWriteError);
+      next.appendFrom(channel, size);
+    } else if (endOffset.get() + size > capacityInBytes) {
+      // maintain the invariant that there are no "partial" writes. A log segment should write the data that it receives
+      // completely and the data that was provided in a single call cannot be written across multiple segments.
+      ensureNext(metrics.overflowWriteError);
+      logger.info("{} : Rolling over writes to {} on write of data of size {} from a channel. "
+              + "Current end offset is {} and capacity is {}", file.getAbsolutePath(), next.getName(), size,
+          getEndOffset(), capacityInBytes);
+      state = State.SEALED;
+      next.appendFrom(channel, size);
+    } else {
+      state = State.ACTIVE;
+      long bytesWritten = 0;
+      while (bytesWritten < size) {
+        bytesWritten += fileChannel.transferFrom(channel, endOffset.get() + bytesWritten, size - bytesWritten);
+      }
+      endOffset.addAndGet(bytesWritten);
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   * <p/>
+   * If {@code offset} + {@code size} exceeds the capacity of the current segment, bytes are written starting from
+   * {@code offset} until capacity is reached. The rest of the bytes upto {@code size} are forwarded to the next segment
+   * (if one exists).
+   * <p/>
+   * The assumption with this function is that the caller is sure of the arguments and understands the implications of
+   * writing at particular offsets and of writing to specific log segment.
+   * <p/>
+   * Using this function does not cause any state changes i.e. this function is expected to be used mostly for
+   * overwriting and not for appending the log.
+   * @param channel The channel from which data needs to be written from.
+   * @param offset The offset at which to write in the underlying write interface.
+   * @param size The amount of data in bytes to be written from the channel.
+   * @throws IllegalArgumentException if {@code offset} < 0 or if there is not enough space for {@code offset } +
+   * {@code size} data and forwarding is not possible
+   * @throws IOException if data could not be written to the file because of I/O errors
+   * @throws UnsupportedOperationException if this LogSegment is intended to be swap space or if it is supposed to free
+   *
+   */
+  @Override
+  public void writeFrom(ReadableByteChannel channel, long offset, long size)
```

Comment:
whatever is the purpose, again I think these semantics will be cleaner if the caller manages the write (so it can ensure that "offset + size" is not greater than capacity and so on).